addons/mail/models/mail_template.py

```diff
@@ -592,91 +592,139 @@ def send_mail(self, res_id, force_send=False, raise_exception=False, email_value
+            else:
+                langs = [False] * len(res_ids_chunk)
+                companies = [self.env.company] * len(res_ids_chunk)
+
+            for values, record, lang, company in zip(values_list, records, langs, companies):
+                values['recipient_ids'] = [(4, pid) for pid in (values.get('partner_ids') or [])]
+                values['attachment_ids'] = [(4, aid) for aid in (values.get('attachment_ids') or [])]
+                values.update(email_values or {})
+
+                # delegate attachments after creation due to ACL check
+                attachments_ids_list.append(values.pop('attachment_ids', []))
+                attachments_list.append(values.pop('attachments', []))
+
+                # add a protection against void email_from
+                if 'email_from' in values and not values.get('email_from'):
+                    values.pop('email_from')
+
+                # encapsulate body
+                if not sending_email_layout_xmlid:
+                    values['body'] = values['body_html']
+                    continue
 
-        mail = self.env['mail.mail'].sudo().create(values)
+                model_lang = model.with_context(lang=lang) if lang else model
+
+                template_ctx = {
+                    # message
+                    'message': self.env['mail.message'].sudo().new(dict(body=values['body_html'], record_name=record.display_name)),
+                    'subtype': self.env['mail.message.subtype'].sudo(),
+                    # record
+                    'model_description': model_lang.display_name,
+                    'record': record,
+                    'record_name': False,
+                    'subtitles': False,
+                    # user / environment
+                    'company': company,
+                    'email_add_signature': False,
+                    'signature': '',
+                    'website_url': '',
+                    # tools
+                    'is_html_empty': is_html_empty,
+                }
+                body = model_lang.env['ir.qweb']._render(sending_email_layout_xmlid, template_ctx, minimal_qcontext=True, raise_if_not_found=False)
+                if not body:
+                    _logger.warning(
+                        'QWeb template %s not found when sending template %s. Sending without layout.',
+                        sending_email_layout_xmlid,
+                        self.name,
+                    )
+                    body = values['body_html']
+
+                values['body_html'] = self.env['mail.render.mixin']._replace_local_links(body)
+                values['body'] = values['body_html']
+
+            mails = self.env['mail.mail'].sudo().create(values_list)
+
+            # manage attachments
+            for mail, attachments, attachment_ids in zip(mails, attachments_list, attachments_ids_list):
+                attachment_ids += [
+                    (0, 0, {
```

Comment:
Well actually new attachments coming from "attachments" (aka: reports) have to be created after the mail. Indeed they are linked to the underlying mail.message, meaning we have to create the mail, then attach those attachments. It is done like that for ACL check: we check acl linked on res_model / res_id of attachments, and currently it is done on the underlying mail.message (as mail.mail is rather technical). All that is a bit strange, but I think purpose is to link attachments to the outgoing mail then remove them, instead of linking them to the record itself. All in all we keep the current behavior.